File: generated_code/GeminiPro/Python/033_SplayTree.py
import sys
from typing import Optional

# A higher recursion limit can be useful for deep tree structures,
# though this splay tree implementation is iterative.
# It is included as a safeguard for potential alternative recursive helpers.
sys.setrecursionlimit(2000)

class SplayTree:
    """
    A complete, self-contained implementation of a Splay Tree that acts as a
    dictionary-like set for integers.

    A splay tree is a self-adjusting binary search tree with the additional
    property that recently accessed elements are quick to access again.
    It performs basic operations such as insertion, look-up, and removal
    in O(log n) amortized time.
    """

    class _Node:
        """A private node class for the SplayTree."""
        def __init__(self, key: int):
            self.key: int = key
            self.parent: Optional['SplayTree._Node'] = None
            self.left: Optional['SplayTree._Node'] = None
            self.right: Optional['SplayTree._Node'] = None

    def __init__(self):
        """Initializes an empty SplayTree."""
        self.root: Optional[SplayTree._Node] = None

    def _left_rotate(self, x: _Node):
        """Performs a left rotation on the subtree rooted at x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x: _Node):
        """Performs a right rotation on the subtree rooted at x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent

        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
            
        if y:
            y.right = x
        x.parent = y

    def _splay(self, x: _Node):
        """
        Performs the splay operation on node x, moving it to the root.
        """
        while x.parent:
            p = x.parent
            g = p.parent
            if not g:  # Zig case
                if x == p.left:
                    self._right_rotate(p)
                else:
                    self._left_rotate(p)
            elif x == p.left and p == g.left:  # Zig-zig case (left-left)
                self._right_rotate(g)
                self._right_rotate(p)
            elif x == p.right and p == g.right:  # Zig-zig case (right-right)
                self._left_rotate(g)
                self._left_rotate(p)
            elif x == p.right and p == g.left:  # Zig-zag case (left-right)
                self._left_rotate(p)
                self._right_rotate(g)
            else:  # Zig-zag case (right-left): x == p.left and p == g.right
                self._right_rotate(p)
                self._left_rotate(g)

    def search(self, key: int) -> bool:
        """
        Searches for a key in the tree.

        This method performs the splaying operation on the accessed node
        (or its parent if the node is not found) to move it to the root.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        current = self.root
        last_node = None
        while current:
            last_node = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:  # key == current.key
                self._splay(current)
                return True
        
        # Key not found, splay the last visited node to the root.
        if last_node:
            self._splay(last_node)
        
        return False

    def insert(self, key: int):
        """
        Inserts a key into the tree.

        If the key already exists, the node with that key is splayed to the root.
        If the key is new, it is inserted and the new node is splayed to the root.

        Args:
            key: The integer key to insert.
        """
        # Handle empty tree case
        if not self.root:
            self.root = self._Node(key)
            return

        # Splay the tree on the key to bring the closest node to the root
        self.search(key)

        # After search, the root is the node with the key or the closest one.
        # If the key is already present, search has already splayed it and we're done.
        if self.root.key == key:
            return

        # Create the new node and insert it by splitting the tree
        new_node = self._Node(key)
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            if new_node.left:
                new_node.left.parent = new_node
            self.root.left = None
        else:  # key > self.root.key
            new_node.left = self.root
            new_node.right = self.root.right
            if new_node.right:
                new_node.right.parent = new_node
            self.root.right = None
        
        self.root.parent = new_node
        self.root = new_node

    def delete(self, key: int):
        """
        Deletes a key from the tree.

        Args:
            key: The integer key to delete.
        """
        if not self.root:
            return

        # Splay the tree on the key. If found, it will be at the root.
        self.search(key)

        # If the key is not at the root after splaying, it wasn't in the tree.
        if self.root.key != key:
            return

        # The node to delete is now at the root.
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # Promote the right subtree to be the new root.
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        elif not right_subtree:
            # Promote the left subtree to be the new root.
            self.root = left_subtree
            if self.root:
                self.root.parent = None
        else:
            # Both subtrees exist. Join them.
            # Disconnect the left subtree to work on it independently.
            left_subtree.parent = None
            
            # Find the maximum element in the left subtree.
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right
            
            # Splay this max element to the root of the (now separate) left subtree.
            # We temporarily set self.root to perform the splay within this subtree.
            self.root = left_subtree
            self._splay(max_node)
            
            # self.root is now the max_node of the original left subtree.
            # By property, it has no right child. Attach the original right subtree here.
            self.root.right = right_subtree
            right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/033_SplayTree.java
/**
 * A self-contained Java class implementing a Splay Tree for an integer set.
 * Splay trees are self-balancing binary search trees with the additional property
 * that recently accessed elements are quick to access again.
 */
public class SplayTree {

    /**
     * Inner class representing a node in the Splay Tree.
     */
    private class Node {
        int key;
        Node left, right;

        /**
         * Constructor for a Node.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty SplayTree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Private Helper Methods for Rotations ---

    /**
     * Performs a right rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    // --- Splay Operation ---

    /**
     * The main splay operation. It brings the node with the given key
     * (or the last accessed node on the search path if the key is not found)
     * to the root of the tree. This is a top-down splay implementation.
     *
     * @param key The key to splay around.
     */
    private void splay(int key) {
        if (root == null) {
            return;
        }

        // A dummy node to simplify linking of left and right subtrees.
        Node header = new Node(0);
        Node leftTreeMax = header;
        Node rightTreeMin = header;
        Node t = root;

        while (true) {
            if (key < t.key) {
                if (t.left == null) break;
                // Zig-Zig (Right-Right) case
                if (key < t.left.key) {
                    t = rightRotate(t);
                    if (t.left == null) break;
                }
                // Link Right: Add current root `t` to the right tree
                rightTreeMin.left = t;
                rightTreeMin = t;
                t = t.left;
            } else if (key > t.key) {
                if (t.right == null) break;
                // Zig-Zig (Left-Left) case
                if (key > t.right.key) {
                    t = leftRotate(t);
                    if (t.right == null) break;
                }
                // Link Left: Add current root `t` to the left tree
                leftTreeMax.right = t;
                leftTreeMax = t;
                t = t.right;
            } else {
                // Key found, break the loop
                break;
            }
        }

        // Reassemble the tree
        leftTreeMax.right = t.left;
        rightTreeMin.left = t.right;
        t.left = header.right;
        t.right = header.left;
        root = t;
    }

    // --- Public API Methods ---

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the tree is splayed on that key.
     *
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, the new node becomes the root.
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree around the key. If the key is present, it becomes the root.
        // If not, the last accessed node on the search path becomes the root.
        splay(key);

        // If the key is already at the root, we are done.
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
            root = newNode;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
            root = newNode;
        }
    }

    /**
     * Searches for a key in the splay tree.
     * This operation performs a splay, moving the accessed node (or the last
     * node on the search path) to the root.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        splay(key);
        // After splaying, if the root's key matches, the key was found.
        if (root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Deletes a key from the splay tree.
     * If the key is found, it is deleted, and the tree is rebalanced.
     * The splay operation is performed on the key before deletion.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree to bring the node to delete (or its parent) to the root.
        splay(key);

        // If the key is not in the tree, the root will be the closest node, but not the key.
        if (root.key != key) {
            return; // Key not found, nothing to delete.
        }

        // Now, the node to delete is the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left child, the right child becomes the new root.
            root = rightSubtree;
        } else {
            // Make the left subtree the main tree.
            root = leftSubtree;
            // Splay on the key we are deleting. Since this key is larger than any element
            // in the left subtree, this will bring the maximum element of the left subtree
            // to its root.
            splay(key);
            // Attach the original right subtree to the new root.
            root.right = rightSubtree;
        }
    }
}